crates/autopilot/src/infra/blockchain/circuit_breaker.rs

```diff
@@ -0,0 +1,131 @@
+use {
+    crate::{
+        domain,
+        infra::blockchain::{
+            contracts::{deployment_address, Addresses},
+            Rpc,
+        },
+    },
+    ethcontract::transaction::TransactionResult,
+    primitive_types::H160,
+    std::{sync::Arc, time::Duration},
+    url::Url,
+};
+
+// Time between Ethereum blocks: 12 seconds
+// In order to fulfill Nyquist theory about sampling rate, we need to at least
+// sample twice the give frequency, in other words, half the period
+const SETTLEMENT_CHECK_SAMPLE_RATE: Duration = Duration::from_secs(6);
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct CircuitBreaker {
+    /// The authenticator contract used for allow-listing solvers to settle.
+    authenticator: contracts::GPv2AllowListAuthentication,
+    /// The safe module that is used to provide special role to EOA.
+    authenticator_role: contracts::Roles,
+    /// The EOA that is allowed to add/remove solvers.
+    authenticator_eoa: ethcontract::Account,
+    /// List of solvers to check
+    solvers: Vec<Url>,
+}
+
+impl CircuitBreaker {
+    /// Spawns a new task in order to monitor the on-chain settlement from the
+    /// solvers
+    pub async fn spawn_monitoring_task(ethrpc: Rpc, addresses: Addresses, solvers: Vec<Url>) {
+        let Rpc { web3, chain, .. } = ethrpc;
+        let Some(authenticator_eoa) = addresses.authenticator_eoa else {
+            return;
+        };
+
+        let address_for = |contract: &ethcontract::Contract, address: Option<H160>| {
+            address
+                .or_else(|| deployment_address(contract, &chain))
+                .unwrap()
+        };
+
+        let settlement = contracts::GPv2Settlement::at(
```

Comment:
This logic already exists when building the `Contracts` struct. I think it would be nicer to just pass a reference to that here to avoid duplicating this logic.

Reply:
@MartinquaXD sure, I can do that, but in that case I need to set the `settlement` before I call both functions.